File: med/cpp/Question1456.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int maxVowels(string s, int k) {
        int ans = 0, cnt = 0;
        for (int i = 0; i < s.length(); i++) {
            // right=i
            // right-left+1=k
            // left=i-k+1
            if (s[i] == 'a' || s[i] == 'e' || s[i] == 'i' || s[i] == 'o' ||
                s[i] == 'u') {
                cnt++;
            }
            if (i < k - 1) {
                continue;
            }
            ans = max(ans, cnt);
            char out = s[i - k + 1];
            if (out == 'a' || out == 'e' || out == 'i' || out == 'o' ||
                out == 'u') {
                cnt--;
            }
        }
        return ans;
    }
};

File: python/med/Question1456.py
class Solution:
    """
    方法一：滑动窗口
        时间复杂度O(n)
        空间复杂度O(1)
    """

    def maxVowels(self, s: str, k: int) -> int:
        left, right = 0, k - 1
        cnt = sum(1 for ch in s[left : right + 1] if ch in "aeiou")
        ans = cnt
        while right < len(s) - 1:
            if s[left] in "aeiou":
                cnt -= 1
            if s[right + 1] in "aeiou":
                cnt += 1
            ans = max(ans, cnt)
            left += 1
            right += 1
        return ans


print(Solution().maxVowels("abciiidef", 3))
